### PR TITLE
feat(polars): initial support for polars backend

### PR DESCRIPTION
**Pull request by ritchie46**
Alright, there is still much to do, but I bounced at this point because I cannot run the tests and therefore not really introspect the types/code.

When I try to run `pytest`, I get this error:

```bash
(venv) (base) ritchie46:~/Downloads/ibis/ibis/backends/polars/tests:  (master)$ pytest conftest.py 
ImportError while loading conftest '/home/ritchie46/Downloads/ibis/ibis/backends/conftest.py'.
../../../__init__.py:6: in <module>
    import ibis.backends.pandas
../../pandas/__init__.py:14: in <module>
    from ibis.backends.base import BaseBackend
../../base/__init__.py:645: in <module>
    sorted(_get_backend_names().difference(("duckdb", "sqlite", "pyspark")))
../../base/__init__.py:638: in _get_backend_names
    entrypoints = importlib.metadata.entry_points()["ibis.backends"]
E   KeyError: 'ibis.backends'

```

Any idea?

### CODE REVIEW

**Comment by cpcloud** [comment]
@ritchie46 Sweet! Looking 👀 now

**Comment by cpcloud** [comment]
@ritchie46 Great starting point.

I'm working on getting the nix builds to pass.

In the meantime I've pushed up a change to the `ibis-backends.yml` workflow to run the polars backend tests.

I'll start reviewing the patch here in a bit.

**Comment by cpcloud** [comment]
Regarding entry points: You'll need to `pip install -e .` for that to work (or `pip install .` but that'll have to be run after every change so I'd recommend against that).

File: ibis/backends/polars/__init__.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import pandas as pd
+import pyarrow as pa
+
+import ibis.common.exceptions as com
+import ibis.expr.analysis as an
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.backends.base import BaseBackend
+from ibis.backends.datafusion.compiler import translate
```
Comment (cpcloud) [inline ibis/backends/polars/__init__.py:18]: I would recommend against importing this, since it requires `datafusion` to be installed.

Comment (ritchie46) [reply ibis/backends/polars/__init__.py:18]: Right, that's a leftover from the copy and paste. Will remove that.

Comment (cpcloud) [inline ibis/backends/polars/__init__.py:18]: I realize from my previous comment it may not be clear what to instead. I think the best option is to copy whatever you can from `ibis/backends/datafusion/compiler.py` and implement the various `translate` rules as needed, in the polars-y way. It might be repetitive, but it removes the datafusion dependency.

Now that we have a second backend in the same spirit as DataFusion we can explore ways to consolidate the rules independent of the specific backend, to avoid repetition.

**Comment by ritchie46** [comment]
> Working on the nix builds, but waiting for the release build to finish. `lto = "fat"` takes a bit of time grinning

You are compiling yourself? Yeap, that takes sometime. :see_no_evil:

File: ibis/backends/polars/__init__.py (hunk 1 above)
Comment (ritchie46) [reply ibis/backends/polars/__init__.py:18]: Yes, I have to still understand this a bit. 

I was thinking of building it around a polars `LazyFrame` which is an abstraction over a computation.

It only gets materialized into a `DataFrame` after a `collect` call. I don't know what's best format for this.

But if I am a bit further I will ask more directed questions.

**Comment by ritchie46** [comment]
@cpcloud Any way I can easily generate the test data?

**Comment by cpcloud** [comment]
> @cpcloud Any way I can easily generate the test data?

Ah, right. I don't think there is. In theory you could take all or most of what's in the DataFusion `conftest.py` file. Let me know if that doesn't work.

**Comment by ritchie46** [comment]
Yeah, I miss the `ibis/ci/ibis-testing-data/functional_alltypes.csv'`.

**Comment by ritchie46** [comment]
@cpcloud I have pushed some more, FYI so that you don't run into merge clashes.

I am stuck on a `ibis.expr.datatypes.core.String is not a valid datatype`.

```
client = <ibis.backends.polars.Backend object at 0x7fabc1cfa520>

    @pytest.fixture(scope='session')
    def alltypes(client):
>       return client.table("functional_alltypes")

tests/conftest.py:69: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
__init__.py:83: in table
    schema = polars_schema_to_ibis(lf.schema)
client.py:54: in polars_schema_to_ibis
    return ibis.schema({name: _polars_to_ibis[dtype] for (name, dtype) in schema.items()})
../../expr/api.py:334: in schema
    return sch.schema(pairs)
../../../venv/lib/python3.9/site-packages/multipledispatch/dispatcher.py:278: in __call__
    return func(*args, **kwargs)
../../expr/schema.py:405: in schema_from_mapping
    return Schema.from_dict(d)
../../expr/schema.py:207: in from_dict
    return cls(names, types)
../../common/grounds.py:25: in __call__
    return cls.__create__(*args, **kwargs)
../../common/grounds.py:111: in __create__
    kwargs = cls.__signature__.validate(*args, **kwargs)
../../common/validators.py:117: in validate
    this[name] = param.validate(value, this=this)
../../common/validators.py:94: in validate
    return self._validator(arg, this=this)
../../../venv/lib/python3.9/site-packages/toolz/functoolz.py:304: in __call__
    return self._partial(*args, **kwargs)
../../common/validators.py:231: in container_of
    return type(inner(item, **kwargs) for item in arg)
../../common/validators.py:231: in <genexpr>
    return type(inner(item, **kwargs) for item in arg)
../../../venv/lib/python3.9/site-packages/toolz/functoolz.py:304: in __call__
    return self._partial(*args, **kwargs)
../../expr/schema.py:47: in datatype
    return dt.dtype(arg)
../../../venv/lib/python3.9/site-packages/multipledispatch/dispatcher.py:278: in __call__
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = <class 'ibis.expr.datatypes.core.String'>, kwargs = {}

    @dtype.register(object)
    def default(value, **kwargs) -> DataType:
>       raise IbisTypeError(f'Value {value!r} is not a valid datatype')
E       ibis.common.exceptions.IbisTypeError: Value <class 'ibis.expr.datatypes.core.String'> is not a valid datatype


```

**Comment by kszucs** [comment]
We need to pass around datatype instances instead of datatype classes. Let me fix that, going to push in a bit.

**Comment by kszucs** [comment]
Things should be a bit more fluent now. The ibis backend test suite can be run for polars using

```console
pytest -sv -m polars ibis/backends/tests/test_generic.py
```

The test errors should clearly indicate the missing operation implementations.

**Comment by ritchie46** [comment]
> Things should be a bit more fluent now. The ibis backend test suite can be run for polars using
> 
> ```
> pytest -sv -m polars ibis/backends/tests/test_generic.py
> ```
> 
> The test errors should clearly indicate the missing operation implementations.

Thanks! Will give it another shot.

**Comment by kszucs** [comment]
If you don't mind I squashed the work so far into a single commit, otherwise the rebase is rather painful.

**Comment by ritchie46** [comment]
> If you don't mind I squashed the work so far into a single commit, otherwise the rebase is rather painful.

Not at all! I am not familiar enough with the ibis internals to make the best steps here, so do a s you please.

Anything missing upstream?

**Comment by kszucs** [comment]
@cpcloud it should be ready. I had to pin the latest polars version which broke the nix builds, could you please fix them?

**Comment by cpcloud** [comment]
@kszucs Yep, looking now.

File: ibis/backends/polars/__init__.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        """Construct a client from a dictionary of polars LazyFrames.
+
+        Parameters
+        ----------
+        _tables
+            Mutable mapping of string table names to polars LazyFrames.
+        """
+        # register dispatchers
+        # from ibis.backends.polars import execution  # noqa: F401
+        # from ibis.backends.polars import udf  # noqa: F401
```
Comment (cpcloud) [inline ibis/backends/polars/__init__.py:41]: Can we delete these lines?

File: ibis/backends/polars/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import ibis
+import ibis.expr.types as ir
+from ibis.backends.tests.base import BackendTest, RoundAwayFromZero
+from ibis.backends.tests.data import array_types
+
+pl = pytest.importorskip("polars")
+
+
+class TestConf(BackendTest, RoundAwayFromZero):
+    bool_is_int = True
```
Comment (cpcloud) [inline ibis/backends/polars/tests/conftest.py:14]: Is this correct? Does polars not support a proper bool type?

File: ibis/backends/polars/tests/test_select.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    )
+    result = expr.execute()
+
+    expected = alltypes_df[
+        (alltypes_df['float_col'] > 0)
+        & (alltypes_df['smallint_col'] == 9)
+        & (alltypes_df['int_col'] < alltypes_df['float_col'] * 2)
+    ]
+
+    BackendTest.assert_frame_equal(result, expected)
```
Comment (cpcloud) [inline ibis/backends/polars/tests/test_select.py:29]: Do we need these tests? This functionality should be covered in the backend suite so we don't have to duplicate tests.

File: ibis/backends/tests/test_generic.py [hunk 5]
```diff
@@ -115,7 +116,12 @@ def test_fillna(backend, alltypes, col):
     [
         param(ibis.coalesce(5, None, 4), 5, id="generic"),
         param(ibis.coalesce(ibis.NA, 4, ibis.NA), 4, id="null_start_end"),
-        param(ibis.coalesce(ibis.NA, ibis.NA, 3.14), 3.14, id="non_null_last"),
+        param(
+            ibis.coalesce(ibis.NA, ibis.NA, 3.14),
+            3.14,
+            id="non_null_last",
+            marks=pytest.mark.broken("polars"),
```
Comment (cpcloud) [inline ibis/backends/tests/test_generic.py:123]: Can you give a `reason=...` here?

File: ibis/backends/tests/test_generic.py [hunk 6]
```diff
@@ -698,7 +707,12 @@ def test_bitwise_columns(backend, con, alltypes, df, op, left_fn, right_fn):
             lambda t: t.int_col,
             id="rshift_col_col",
         ),
-        param(rshift, lambda _: 3, lambda t: t.int_col, id="rshift_scalar_col"),
+        param(
+            rshift,
+            lambda _: 3,
+            lambda t: t.int_col,
+            id="rshift_scalar_col",
```
Comment (cpcloud) [inline ibis/backends/tests/test_generic.py:714]: This line looks the same as before, any reason to change the formatting?

File: ibis/backends/tests/test_numeric.py [hunk 7]
```diff
@@ -429,7 +429,7 @@ def test_floating_mod(backend, alltypes, df):
         'smallint_col',
         'int_col',
         'bigint_col',
-        'float_col',
+        pytest.param('float_col', marks=pytest.mark.broken("polars", strict=False)),
```
Comment (cpcloud) [inline ibis/backends/tests/test_numeric.py:432]: Can you give a `reason=...` here? Why disable `strict`?

File: ibis/backends/tests/test_temporal.py [hunk 8]
```diff
@@ -854,8 +868,8 @@ def test_integer_cast_to_timestamp(backend, alltypes, df):
     ),
 )
 @pytest.mark.notimpl(
-    ["datafusion"],
-    reason="DataFusion backend assumes ns resolution timestamps",
+    ["datafusion", "polars"],
+    reason="Arrow backends assumes ns resolution timestamps",
```
Comment (cpcloud) [inline ibis/backends/tests/test_temporal.py:872]:
```suggestion
    reason="Arrow backends assume a ns resolution timestamps",
```

**Review by cpcloud — CHANGES_REQUESTED** [review]
Looking really good!

A few small things to address, but after that this is good to go.

File: ibis/backends/polars/__init__.py (hunk 2 above)
Comment (kszucs) [reply ibis/backends/polars/__init__.py:41]: Sure

File: ibis/backends/polars/tests/conftest.py (hunk 3 above)
Comment (kszucs) [reply ibis/backends/polars/tests/conftest.py:14]: It dies, fixing.

File: ibis/backends/polars/tests/test_select.py (hunk 4 above)
Comment (kszucs) [reply ibis/backends/polars/tests/test_select.py:29]: Removing

File: ibis/backends/tests/test_generic.py (hunk 5 above)
Comment (kszucs) [reply ibis/backends/tests/test_generic.py:123]: Yes

File: ibis/backends/tests/test_generic.py (hunk 6 above)
Comment (kszucs) [reply ibis/backends/tests/test_generic.py:714]: Restoring the formatting

File: ibis/backends/tests/test_numeric.py (hunk 7 above)
Comment (kszucs) [reply ibis/backends/tests/test_numeric.py:432]: Yes

**Review by cpcloud — APPROVED** [review]
Other than the failing docs job this LGTM!

File: ibis/backends/polars/tests/conftest.py (hunk 3 above)
Comment (kszucs) [reply ibis/backends/polars/tests/conftest.py:14]: Updated.

File: ibis/backends/polars/__init__.py (hunk 2 above)
Comment (kszucs) [reply ibis/backends/polars/__init__.py:41]: Removed.

File: ibis/backends/polars/tests/test_select.py (hunk 4 above)
Comment (kszucs) [reply ibis/backends/polars/tests/test_select.py:29]: Removed.

File: ibis/backends/tests/test_generic.py (hunk 6 above)
Comment (kszucs) [reply ibis/backends/tests/test_generic.py:714]: Restored.

File: ibis/backends/tests/test_temporal.py (hunk 8 above)
Comment (kszucs) [reply ibis/backends/tests/test_temporal.py:872]: Updated.

File: ibis/backends/tests/test_generic.py (hunk 5 above)
Comment (kszucs) [reply ibis/backends/tests/test_generic.py:123]: Added.

**Comment by ritchie46** [comment]
Thanks for all the help on this!
